### PR TITLE
WELD-2334 Document WeldInstance, WeldAfterBeanDiscovery and WeldInvoc…

### PR DESCRIPTION
**Pull request by manovotn**
…ationContext.

### CODE REVIEW

File: docs/reference/src/main/asciidoc/extend.asciidoc [hunk 1]
```diff
@@ -138,13 +138,43 @@ Container lifecycle event observer methods may inject a `BeanManager`:
...
+The interceptor is created automatically, once the methods exits and the configurator instance is not reusable.
+But if you need to create several interceptors, you can simply request several configurator instances.
+Here is a code snippet to demonstrate the idea:
+
+[source.JAVA, java]
+------------------------------------------------------------------------------------------------------------------------------------
+public void afterBeanDiscovery(@Observes WeldAfterBeanDiscovery event) {
+
+        // type level interceptor
+        event.addInterceptor().interceptWithMetadata(InterceptionType.AROUND_INVOKE, (invocationContext, fooBean) -> {
```
Comment (mkouba) [inline docs/reference/src/main/asciidoc/extend.asciidoc]: If you don't use the intercepted bean (metadata) you should not use `interceptWithMetadata()` but `intercept()`.

File: docs/reference/src/main/asciidoc/extend.asciidoc [hunk 2]
```diff
@@ -138,13 +138,43 @@ Container lifecycle event observer methods may inject a `BeanManager`:
...
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+            return null;
+        }).priority(2500).addBinding(MyTypeBinding.MyTypeBindingLiteral.INSTANCE);
+}
+------------------------------------------------------------------------------------------------------------------------------------
+
+As you might have noticed, the sample shows an interception with metadata, meaning the the body method is represented as `java.util.function.BiFunction` where the second argument is an injection point.
+Obviously, there is also a method for simple interception
```
Comment (mkouba) [inline docs/reference/src/main/asciidoc/extend.asciidoc]: I would rather show an example for simple interception...

File: docs/reference/src/main/asciidoc/extend.asciidoc (hunk 1 above)
Comment (manovotn) [inline docs/reference/src/main/asciidoc/extend.asciidoc]: Good point, I meant to add `fooBean` as a parameter to `getAnswerToLifeTheUniverseAndEverything` but forgot :)

File: docs/reference/src/main/asciidoc/injection.asciidoc [hunk 3]
```diff
@@ -564,6 +564,83 @@ Since CDI 2.0, most annotations from `javax.enterprise` package have their `Anno
...
+            if (Dependent.class.equals(handler.getBean().getScope()) {
+                // Destroy only dependent processors
+                handler.destroy();
+            }
+        }
+  }
+}
+----------------------------------------------------------------------------------
+
+And the last one `isResolvable()` is just a convenient method - a replacement for `!isUnsatisfied() && !isAmbiguous()` which is the expression most users are interested in:
```
Comment (mkouba) [inline docs/reference/src/main/asciidoc/injection.asciidoc]: This is only Weld-specific in 2.4.

**Comment by manovotn** [comment]
@mkouba incorporated the changes you suggested. How do you like it now?

File: docs/reference/src/main/asciidoc/extend.asciidoc [hunk 4]
```diff
@@ -138,13 +138,44 @@ Container lifecycle event observer methods may inject a `BeanManager`:
...
+[[_configurators]]
 ==== Configurators
 
 CDI 2.0 introduced the Configurators API - a new way to easily configure some parts of the SPI during container lifecycle event notification.
-E.g. to add a qualifier to a bean an extension can observe `ProcessBeanAttributes`, then obtain a configurator instance through `ProcessBeanAttributes.configureBeanAttributes()` and finally use `BeanAttributesConfigurator.addQualifier(Annotation)`.
+E.g. to add a qualifier to a bean, an extension can observe `ProcessBeanAttributes`, then obtain a configurator instance through `ProcessBeanAttributes.configureBeanAttributes()` and finally use `BeanAttributesConfigurator.addQualifier(Annotation)`.
 No need to wrap/delegate to the original `BeanAttributes`.
 See also chapter http://docs.jboss.org/cdi/spec/2.0-PRD/cdi-spec.html#configurators[Configurators interfaces] of the CDI specification.
 
+==== Weld-enriched Container lifecycle events
```
Comment (mkouba) [inline docs/reference/src/main/asciidoc/extend.asciidoc]: Weld-enriched **c**~C~ontainer lifecycle events?

File: docs/reference/src/main/asciidoc/extend.asciidoc [hunk 5]
```diff
@@ -138,13 +138,44 @@ Container lifecycle event observer methods may inject a `BeanManager`:
...
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+            return null;
+        }).priority(2500).addBinding(MyTypeBinding.MyTypeBindingLiteral.INSTANCE);
+}
+------------------------------------------------------------------------------------------------------------------------------------
+
+The sample presents a simple interception use case where you supply `java.util.function.Function` as the interceptor body method.
+For more comples cases, you may also choose to use `interceptWithMetadata` method which accepts `java.util.function.BiFunction` instead.
```
Comment (mkouba) [inline docs/reference/src/main/asciidoc/extend.asciidoc]: comples -> complex

File: docs/reference/src/main/asciidoc/extend.asciidoc [hunk 6]
```diff
@@ -138,13 +138,44 @@ Container lifecycle event observer methods may inject a `BeanManager`:
...
+The second parameter of the `BiFunction` is an injection point which will be automatically resolved by Weld.
```
Comment (mkouba) [inline docs/reference/src/main/asciidoc/extend.asciidoc]: > ...an injection point which will be automatically resolved by Weld...

Not really. The second param emulates `@Inject @Intercepted Bean<?>`, see also http://docs.jboss.org/cdi/spec/1.2/cdi-spec.html#bean_metadata.

**Comment by mkouba** [comment]
Otherwise looks good ;-)

**Pull request merged**